### PR TITLE
Fix the two bugs in assembly with array variables

### LINKED ISSUE #23123: Bugs in assembling Jacobian with array variables
## Bug Description
These bugs were discovered by Namjae when he tried to use a custom Griffin executioner which uses `ComputeJacobianBlocksThread` to assemble the block-diagonal matrix. `ComputeJacobianBlocksThread` calls `FEProblemBase::addJacobianBlockTags`, which calls `Assembly::addJacobianBlock`, by passing in a libMesh variable number for any component of array variables. Current code in `Assembly::addJacobianBlock` assumes the variable number passed in is the libMesh variable number of the 0-th component. Another bug is in `Assembly::saveFullLocalArrayJacobian` which is called through either `ComputeJacobianBlocksThread` or `ComputeFullJacobianThread` through kernel/bc `computeOffDiagJacobian` function. The case with component block diagonal is handled wrongly.

## Steps to Reproduce
After fixing those two bugs, the test in Griffin can run properly. I will push the fix in first and see whether I can come up with a MOOSE test for the fix.

## Impact
Fixing bugs that can show up in corner cases.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #23123.

### CODE REVIEW

**Comment by YaqiWang** [comment]
This should pass all the tests. I will add a small test later.

**Comment by lindsayad** [comment]
So we don't have a way to do better than the 36% coverage?

**Comment by YaqiWang** [comment]
I thought I would got 100% coverage with the test lol. Let me check the report.

**Comment by YaqiWang** [comment]
Looks like I just modified codes that was not covered before and those modifications are marked uncovered now. We probably have tests for some of the codes in applications.

**Comment by lindsayad** [comment]
I'm just worried if you say that you are fixing two bugs but the bug fixes are not covered, how do we know we will not regress or that the bugs are fixed? It's one thing not to cover a new feature; it's another thing to say there are bugs and that you're fixing them and not to cover what are reported as bugs

**Comment by YaqiWang** [comment]
The two bugs we notices are covered, one in `Assembly.h` another in `Assembly::addJacobianBlock`. While working on the later, I notices other `addJacobianBlockxxx`s had similar issue as `Assembly::addJacobianBlock`. The uncovered lines are in those functions. Supposedly if we have tests calling those functions, those lines will be covered as well, right?

**Comment by lindsayad** [comment]
Of course. And clearly those tests don't exist. You are implying that it shouldn't be your responsibility to add tests for code that wasn't yours that you fixed. And that's fine. You'll just have to wait a bit longer on this PR then for me to come up with tests for those uncovered lines

**Comment by lindsayad** [comment]
Or if you are in a big hurry, you can split the PR into the fixes that are relevant to Griffin which it sounds like you already have test coverage for and the fixes for which you do not yet have coverage for. And I can add tests for the latter

**Comment by YaqiWang** [comment]
Yes, you got me right ;-) I might be able to switch the test to DG to have more coverage. But those nonlocal stuff will still not be covered probably.

**Comment by YaqiWang** [comment]
But it is very strange to me that `Assembly::prepareBlock` must be used in existing tests, yet the changes in it shows not covered. Any idea on this?

**Comment by YaqiWang** [comment]
Switched to use PBP and hopefully that will increase the coverage.

File: framework/include/base/Assembly.h [hunk 1]
```diff
@@ -1629,13 +1629,20 @@ class Assembly
...
     {
-      j = saved_j;
-      for (unsigned int l = 0; l < v.cols(); ++l, j += pace)
-        ke(i, j) += v(k, l);
+      for (unsigned int k = 0; k < v.rows(); ++k, i += ntest)
+        ke(i, j) += v(k, k);
+    }
+    else
+    {
+      unsigned int saved_j = j;
```
Comment (lindsayad) [inline framework/include/base/Assembly.h:1639]:
```suggestion
      const unsigned int saved_j = j;
```

File: framework/src/base/Assembly.C [hunk 2]
```diff
@@ -2837,19 +2837,23 @@ Assembly::prepareBlock(unsigned int ivar,
...
 {
-  unsigned int icount = _sys.getVariable(_tid, ivar).count();
-  unsigned int jcount = _sys.getVariable(_tid, jvar).count();
-  if (ivar == jvar && _component_block_diagonal[ivar])
+  auto & iv = _sys.getVariable(_tid, ivar);
+  auto & jv = _sys.getVariable(_tid, jvar);
+  unsigned int ivn = iv.number();
+  unsigned int jvn = jv.number();
+  unsigned int icount = iv.count();
+  unsigned int jcount = jv.count();
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:2845]:
```suggestion
  const unsigned int ivn = iv.number();
  const unsigned int jvn = jv.number();
  const unsigned int icount = iv.count();
  const unsigned int jcount = jv.count();
```

File: framework/src/base/Assembly.C [hunk 3]
```diff
@@ -2858,19 +2862,23 @@ Assembly::prepareBlockNonlocal(unsigned int ivar,
...
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:2870]:
```suggestion
  const unsigned int ivn = iv.number();
  const unsigned int jvn = jv.number();
  const unsigned int icount = iv.count();
  const unsigned int jcount = jv.count();
```

File: framework/src/base/Assembly.C [hunk 4]
```diff
@@ -4222,19 +4230,22 @@ Assembly::addJacobianBlock(SparseMatrix<Number> & jacobian,
   if (!(*_cm)(ivar, jvar))
     return;
 
-  DenseMatrix<Number> & ke = jacobianBlock(ivar, jvar, tag);
   auto & iv = _sys.getVariable(_tid, ivar);
   auto & jv = _sys.getVariable(_tid, jvar);
   auto & scaling_factor = iv.arrayScalingFactor();
 
+  unsigned int ivn = iv.number();
+  unsigned int jvn = jv.number();
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:4238]:
```suggestion
  const unsigned int ivn = iv.number();
  const unsigned int jvn = jv.number();
```

File: framework/src/base/Assembly.C [hunk 5]
```diff
@@ -4222,19 +4230,22 @@ Assembly::addJacobianBlock(SparseMatrix<Number> & jacobian,
...
   // It is guaranteed by design iv.number <= ivar since iv is obtained
   // through SystemBase::getVariable with ivar.
   // Most of times ivar will just be equal to iv.number except for array variables,
   // where ivar could be a number for a component of an array variable but calling
   // getVariable will return the array variable that has the number of the 0th component.
   // It is the same for jvar.
-  unsigned int i = ivar - iv.number();
-  unsigned int j = jvar - jv.number();
+  unsigned int i = ivar - ivn;
+  unsigned int j = jvar - jvn;
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:4248]:
```suggestion
  const unsigned int i = ivar - ivn;
  const unsigned int j = jvar - jvn;
```

File: framework/src/base/Assembly.C [hunk 6]
```diff
@@ -4275,19 +4286,22 @@ Assembly::addJacobianBlockNonlocal(SparseMatrix<Number> & jacobian,
   if (!(*_cm)(ivar, jvar))
     return;
 
-  DenseMatrix<Number> & keg = jacobianBlockNonlocal(ivar, jvar);
   auto & iv = _sys.getVariable(_tid, ivar);
   auto & jv = _sys.getVariable(_tid, jvar);
   auto & scaling_factor = iv.arrayScalingFactor();
 
+  unsigned int ivn = iv.number();
+  unsigned int jvn = jv.number();
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:4294]:
```suggestion
  const unsigned int ivn = iv.number();
  const unsigned int jvn = jv.number();
```

File: framework/src/base/Assembly.C [hunk 7]
```diff
@@ -4275,19 +4286,22 @@ Assembly::addJacobianBlockNonlocal(SparseMatrix<Number> & jacobian,
...
   // It is guaranteed by design iv.number <= ivar since iv is obtained
   // through SystemBase::getVariable with ivar.
   // Most of times ivar will just be equal to iv.number except for array variables,
   // where ivar could be a number for a component of an array variable but calling
   // getVariable will return the array variable that has the number of the 0th component.
   // It is the same for jvar.
-  unsigned int i = ivar - iv.number();
-  unsigned int j = jvar - jv.number();
+  unsigned int i = ivar - ivn;
+  unsigned int j = jvar - jvn;
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:4304]:
```suggestion
  const unsigned int i = ivar - ivn;
  const unsigned int j = jvar - jvn;
```

File: framework/src/base/Assembly.C [hunk 8]
```diff
@@ -4326,22 +4340,24 @@ Assembly::addJacobianNeighbor(SparseMatrix<Number> & jacobian,
...
-  DenseMatrix<Number> & ken = jacobianBlockNeighbor(Moose::ElementNeighbor, ivar, jvar);
-  DenseMatrix<Number> & kne = jacobianBlockNeighbor(Moose::NeighborElement, ivar, jvar);
-  DenseMatrix<Number> & knn = jacobianBlockNeighbor(Moose::NeighborNeighbor, ivar, jvar);
-
   auto & iv = _sys.getVariable(_tid, ivar);
   auto & jv = _sys.getVariable(_tid, jvar);
   auto & scaling_factor = iv.arrayScalingFactor();
 
+  unsigned int ivn = iv.number();
+  unsigned int jvn = jv.number();
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:4348]:
```suggestion
  const unsigned int ivn = iv.number();
  const unsigned int jvn = jv.number();
```

File: framework/src/base/Assembly.C [hunk 9]
```diff
@@ -4326,22 +4340,24 @@ Assembly::addJacobianNeighbor(SparseMatrix<Number> & jacobian,
...
   // It is guaranteed by design iv.number <= ivar since iv is obtained
   // through SystemBase::getVariable with ivar.
   // Most of times ivar will just be equal to iv.number except for array variables,
   // where ivar could be a number for a component of an array variable but calling
   // getVariable will return the array variable that has the number of the 0th component.
   // It is the same for jvar.
-  unsigned int i = ivar - iv.number();
-  unsigned int j = jvar - jv.number();
+  unsigned int i = ivar - ivn;
+  unsigned int j = jvar - jvn;
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:4361]:
```suggestion
  const unsigned int i = ivar - ivn;
  const unsigned int j = jvar - jvn;
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
@YaqiWang if you really want the new `view_matrix` option in PBP, we can strip off that last commit. But if we do, let's create a `_view_matrix` const boolean class member so we don't have to call `getParam` every time the operator might change

**Comment by YaqiWang** [comment]
> @YaqiWang if you really want the new `view_matrix` option in PBP, we can strip off that last commit. But if we do, let's create a `_view_matrix` const boolean class member so we don't have to call `getParam` every time the operator might change

Not really. I do not know we can use `-mat_view` to show the matrices.

**Review by lindsayad — APPROVED** [review]

**Comment by YaqiWang** [comment]
Thanks @lindsayad !
